transaction/std/src/memo_builder/gift_code_memo_builder.rs

```diff
@@ -0,0 +1,181 @@
+/// | -----------     | -----------                                       |
+/// | 0x0002          | Gift Code Sender Memo                             |
+/// | 0x0201          | Gift Code Funding Memo                            |
+/// | 0x0202          | Gift Code Cancellation Memo                       |
+#[derive(Clone, Debug)]
+pub enum GiftCodeMemoType {
+    /// Gift Code Sender Memo 0x0002
+    Sender,
+    /// Gift Code Funding Memo 0x0201
+    Funding,
+    /// Gift Code Cancellation Memo 0x0202
+    Cancellation,
+}
+
+/// This memo builder writes gift code memos to the change address which can be
+/// used by apps that offer gift codes as feature to track the status of gift
+/// codes as well as certain contextual information about them. This builder
+/// writes 3 possible gift code memo types. Only one memo which can be specified
+/// at the creation of the memo builder. It is worth noting that all memos
+/// written by this memo builder will go to the reserved change address and
+/// does not write any destination memos. For normal recoverable transaction
+/// history (RTH), the RTH memo builder should be used in conjuction with this
+/// memo builder.
+#[derive(Clone, Debug)]
+pub struct GiftCodeMemoBuilder {
+    // The gift code memo type we'll be writing
+    gift_code_memo_type: GiftCodeMemoType,
+    // The utf-8 formatted note that appears on the change address when a gift code is
+    // or cancelled
+    note: &'static str,
+    // TxOut Public Key of the gift code if it it's a funding memo
+    gift_code_tx_out_public_key: Option<RistrettoPublic>,
+    // Global TxOut index of the gift code if it's a cancellation memo
+    gift_code_tx_out_global_index: Option<u64>,
+    // Whether or not to enable gift code change memos
+    gift_code_change_memo_enabled: bool,
+    // Whether we've already written the change memo
+    wrote_change_memo: bool,
+}
+
+impl GiftCodeMemoBuilder {
+    /// Create a new gift code memo builder specifying the memo type
+    pub fn new(gift_code_memo_type: GiftCodeMemoType) -> Self {
+        Self {
+            gift_code_memo_type,
+            note: "",
+            gift_code_tx_out_public_key: None,
+            gift_code_tx_out_global_index: None,
+            gift_code_change_memo_enabled: true,
+            wrote_change_memo: false,
+        }
+    }
+    /// A note in the form of null terminated utf-8 bytes can be applied to
+    /// gift code funding or sender memos. These notes end up in the change
+    /// outputs and indicate what the memo was for. Set this memo here. This
+    /// method will enforce the 60 and 64 byte length for Funding Memos and
+    /// Sender Memos respectively. If the memo type is a cancellation memo
+    /// and a note is set, any value set here will be ignored as the
+    /// cancellation memo does not include a note.
+    pub fn set_note(&mut self, note: &'static str) -> Result<(), MemoError> {
```

Comment:
This shouldn't be `&'static str`, that limits it to strings embedded in the code.

Reply:
This is my "learning Rust" curve showing. Thanks for pointing out that embedded doesn't cover all.